lib/maths/CTreeShapFeatureImportance.cc

```diff
@@ -0,0 +1,258 @@
+                                 std::size_t offset) {
+    numberFeatures = (numberFeatures != -1) ? numberFeatures : frame.numberColumns();
+    std::vector<std::size_t> maxDepthVec;
+    maxDepthVec.reserve(m_Trees.size());
+    for (auto& tree : m_Trees) {
+        auto samplesPerNode =
+            CTreeShapFeatureImportance::samplesPerNode(tree, frame, encoder, 0);
+        std::size_t maxDepth =
+            CTreeShapFeatureImportance::updateNodeValues(tree, 0, samplesPerNode, 0);
+        maxDepthVec.push_back(maxDepth);
+        m_SamplesPerNode.emplace_back(std::move(samplesPerNode));
+    }
+
+    auto result = frame.writeColumns(
+        m_NumberThreads, 0, frame.numberRows(),
+        core::bindRetrievableState(
+            [&](auto& state, TRowItr beginRows, TRowItr endRows) {
+                for (auto row = beginRows; row != endRows; ++row) {
+                    auto encodedRow{encoder.encode(*row)};
+                    auto& phiSum = state;
+                    for (int i = 0; i < m_Trees.size(); ++i) {
+                        //                        phi[frame.numberColumns()] += m_Trees[i][0].value();
+                        SPath path(maxDepthVec[i] + 1);
+                        CTreeShapFeatureImportance::shapRecursive(
+                            m_Trees[i], m_SamplesPerNode[i], encoder,
+                            encodedRow, path, 0, 1.0, 1.0, -1, offset, row);
+                        i += 0;
+                    }
+                    for (int j = 0; j < numberFeatures; ++j) {
+                        row->writeColumn(offset + j,
+                                         (*row)[offset + j] / m_Trees.size());
+                        phiSum[j] += std::fabs((*row)[offset + j]);
+                    }
+
+                    //                    phiVec.emplace_back(std::move(phi));
+                }
+            },
+            TDoubleVec(numberFeatures, 0)));
+
+    auto& state = result.first;
+    TDoubleVec phiSum{std::move(state[0].s_FunctionState)};
+    for (int i = 1; i < state.size(); ++i) {
+        auto& otherPhiSum = state[i].s_FunctionState;
+        std::transform(phiSum.begin(), phiSum.end(), otherPhiSum.begin(),
+                       phiSum.begin(), std::plus<double>());
+    }
+
+    return phiSum;
+}
+
+CTreeShapFeatureImportance::TDoubleVec
+CTreeShapFeatureImportance::samplesPerNode(const TTree& tree,
+                                           const core::CDataFrame& frame,
+                                           const CDataFrameCategoryEncoder& encoder,
+                                           std::size_t numThreads) {
+    auto result = frame.readRows(
+        numThreads, 0, frame.numberRows(),
+        core::bindRetrievableState(
+            [&](TDoubleVec& state, TRowItr beginRows, TRowItr endRows) {
+                for (auto row = beginRows; row != endRows; ++row) {
+                    auto encodedRow{encoder.encode(*row)};
+                    auto node{tree[0]};
```

Comment:
I think we should make this type `const CBoostedTreeNode*` otherwise you are copying the entire node as you traverse the tree for every single row (and knock on changes below).
```suggestion
                    const CBoostedTreeNode* node{&tree[0]};
```

Reply:
fixed